### ISSUE TITLE
distribute_dofs() throwing error when called on a externally read parallel triangulation with periodic bnd

### ISSUE THREAD

**Issue by dsambit** [dealii/dealii#5642]
[debugFiles.zip](https://github.com/dealii/dealii/files/1568230/debugFiles.zip)
In the above I have attached a minimal example which reads an external uniform mesh (mesh.inp). It is of cuboidal shape with dimensions (7.6 x 7.6 x 7.6), origin at the center and has 1000 hex elements. Next I apply periodic boundary conditions and call distrubute_dofs for linear shape functions- I get the following error

An error occurred in line <3873> of file </home/vikramg/DFT-FE-softwares/softwareCentos/dealii-patched/dealii/source/dofs/dof_handler_policy.cc> in function
    dealii::internal::DoFHandler::NumberCache dealii::internal::DoFHandler::Policy::ParallelDistributed<DoFHandlerType>::distribute_dofs() const [with DoFHandlerType = dealii::DoFHandler<3, 3>]
The violated condition was: 
    cell->user_flag_set() == false
Additional information: 
    This exception -- which is used in many places in the library -- usually indicates that some condition which the author of the code thought must be satisfied at a certain point in an algorithm, is not fulfilled. An example would be that the first part of an algorithm sorts elements of an array in ascending order, and a second part of the algorithm later encounters an element that is not larger than the previous one.

There is usually not very much you can do if you encounter such an exception since it indicates an error in deal.II, not in your own program. Try to come up with the smallest possible program that still demonstrates the error and contact the deal.II mailing lists with it to obtain help.

Stacktrace:
-----------
```
#0  /home/vikramg/DFT-FE-softwares/softwareCentos/dealii-patched/install_intel_17.0.1_bugfix1/lib/libdeal_II.g.so.9.0.0-pre: dealii::internal::DoFHandler::Policy::ParallelDistributed<dealii::DoFHandler<3, 3> >::distribute_dofs() const
#1  /home/vikramg/DFT-FE-softwares/softwareCentos/dealii-patched/install_intel_17.0.1_bugfix1/lib/libdeal_II.g.so.9.0.0-pre: dealii::DoFHandler<3, 3>::distribute_dofs(dealii::FiniteElement<3, 3> const&)
#2  ./minimalExampleBugReproduceNewVersionReadMesh: main

The above error only occurs when I
a)  add  periodic boundary conditions to the triangulation
b) parallel run (I specifically ran on 16 processors)
c) read an external mesh
```

**Comment by bangerth** [comment]
@dsambit -- thanks! 

I'm trying to run this on my laptop, but it takes a surprisingly long time -- longer than I want that to take on my laptop. It's only 1000 cells, and so should really not take any kind of time at all. How long does it run on your machine before it crashes?

If you wanted to try a couple of things: Is it possible to reproduce this with fewer than the 1000 cells, and/or with fewer than 16 processors?

**Comment by dsambit** [comment]
I does take a long time. For me it crashes after 6 and a half minutes. I am now trying to generate a smaller mesh using CUBIT-about 100-200 cells and reproduce the error.

**Comment by bangerth** [comment]
Appreciated. It makes no sense to me that it should take that long -- there must be some wayward loop somewhere in which we get stuck. I'm afraid this is going to be a rather difficult bug to track down. Any simplification you can offer would be much appreciated!

(Side remark: you can probably also remove the code that sets manifold -- you don't do refinement, so it has no effect.)

**Comment by dsambit** [comment]
I have now reproduced the bug on a 64 element uniform mesh. The domain dimensions are still the same. Good thing is now it takes a few seconds for the error to throw. However, it throws the error only when I run on 16 processors but not on anything less. I have also removed the code that sets manifold. The files are attached below
[debugFiles2.zip](https://github.com/dealii/dealii/files/1568521/debugFiles2.zip)
